hello-data-sidecars/hello-data-sidecar-superset/src/main/java/ch/bedag/dap/hellodata/sidecars/superset/service/dashboard/UploadDashboardsFileListener.java

```diff
@@ -0,0 +1,181 @@
+                    buildChunks(dashboardUpload.getBinaryFileId(), dashboardUpload.getFilename(), dashboardUpload.getChunkNumber(), dashboardUpload.getFileSize(),
+                                contentFile.toPath());
+                } else {
+                    log.debug("Saved chunk, waiting for another one {}", dashboardUpload.getChunkNumber());
+                    ackMessage(msg);
+                    return;
+                }
+                supersetClient.importDashboard(contentFile, new JsonPrimitive("{}"), true);
+                log.debug("\t-=-=-=-= received message from the superset: {}", new String(msg.getData()));
+                ackMessage(msg);
+            } catch (URISyntaxException | IOException | RuntimeException e) {
+                log.error("Error uploading dashboards", e);
+                natsConnection.publish(msg.getReplyTo(), e.getMessage().getBytes(StandardCharsets.UTF_8));
+            }
+        });
+        dispatcher.subscribe(supersetSidecarSubject);
+    }
+
+    private void ackMessage(Message msg) {
+        natsConnection.publish(msg.getReplyTo(), "OK".getBytes(StandardCharsets.UTF_8));
+        msg.ack();
+    }
+
+    public void saveChunk(DashboardUpload chunk) throws IOException {
+        Path uploadFolderPath = createTempFolder(chunk.getBinaryFileId());
+        Path path = Paths.get(uploadFolderPath.toString(), chunk.getChunkNumber() + CHUNK_SUFFIX);
+        Files.write(path, chunk.getContent());
+    }
+
+    private Path createTempFolder(String filename) {
+        Path uploadFolderPath = getUploadFolderPath(filename);
+        if (!Files.exists(uploadFolderPath)) {
+            File file = uploadFolderPath.toFile();
+            boolean created = file.mkdirs();
+            log.debug("File {} created: {}", file.toPath(), created);
+        }
+        return uploadFolderPath;
+    }
+
+    /**
+     * Regex pattern replace all but numbers, letters, dashes and underscores.
+     */
+    private Path getUploadFolderPath(String filename) {
+        String uploadFolder = filename.replaceAll(FOLDER_NAMES_REGEX_PATTERN, "");
+        return Paths.get(System.getProperty("java.io.tmpdir"), "dashboards_upload", uploadFolder);
+    }
+
+    private void buildChunks(String binaryFileId, String filename, long totalChunks, long fileSize, Path destinationPath) throws IOException {
+        Path chunksFolderPath = getUploadFolderPath(binaryFileId);
+
+        if (!Files.exists(chunksFolderPath)) {
+            throw new RuntimeException("No chunks were found for filename : " + filename);
+        }
+
+        List<File> chunks = listChunks(chunksFolderPath);
+        if (chunks.isEmpty() || chunks.size() != totalChunks || validateChunkSize(fileSize, chunks)) {
+            throw new RuntimeException("The file is not supported. Upload canceled");
+        }
+        writeChunksToFile(destinationPath, chunks);
+        deleteTempBinaryFileData(binaryFileId);
```

Comment:
maybe this delete-method could be moved into the try/catch/finally of the listenForRequests-method, so it gets called even in case of an exception.

Reply:
Sure, no problem. But I need the binary file ID to be able to be deleted. So there might be a situation where the folder will remain on exception....